File: src/main/java/com/server/computerscience/chatbot/service/implement/ChatCacheService.java
package com.server.computerscience.chatbot.service.implement;

import java.util.LinkedList;
import java.util.List;

import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import com.server.computerscience.chatbot.dto.request.ChatMessageDto;

@Service
public class ChatCacheService {
	private static final String CHAT_MESSAGE = "chatMessages";
	private static final String CHAT_USED_CHANCE = "chatRemainChance";
	private static final int MAX_CHAT_CHANCE = 20;
	private static final int MAX_MESSAGES_SIZE = 10;

	/**
	 * 이미 저장되어있는 대화목록이 있을 경우
	 * 아니라면 그대로 반환
	 */
	@Cacheable(value = CHAT_MESSAGE, key = "#userId")
	public List<ChatMessageDto> getChatMessages(String userId) {
		return new LinkedList<>();
	}

	@CachePut(value = CHAT_MESSAGE, key = "#userId")
	public List<ChatMessageDto> saveChatMessage(String userId, ChatMessageDto chatMessageDto) {
		List<ChatMessageDto> chatMessages = getChatMessages(userId);
		chatMessages.add(chatMessageDto);
		if (chatMessages.size() > MAX_MESSAGES_SIZE) {
			chatMessages.remove(0);
		}
		return chatMessages;
	}

	/**
	 * 특정 사용자의 메시지 캐시를 삭제하는 메서드
	 */
	@CacheEvict(value = CHAT_MESSAGE, key = "#userId")
	public void clearUserMessages(String userId) {
		// 캐시에서 삭제할 내용이 없으면 자동으로 무시됩니다
	}

	/**
	 * 챗봇 이용 횟수 제한
	 */
	public boolean canChat(String userId) {
		// 남은 기회가 0 이하인 경우 false 반환
		return getUsedChance(userId) < MAX_CHAT_CHANCE;
	}

	// 남은 이용 횟수를 가져오는 메서드
	@Cacheable(value = CHAT_USED_CHANCE, key = "#userId")
	public int getUsedChance(String userId) {
		return 0;
	}

	// 사용 횟수를 증가시키는 메서드
	@CachePut(value = CHAT_USED_CHANCE, key = "#userId")
	public int increaseUsedChance(String userId) {
		int currentUsedChance = getUsedChance(userId);
		return currentUsedChance + 1; // 현재 사용 횟수에 1 추가
	}

}

File: src/main/java/com/server/computerscience/chatbot/service/implement/ChatCacheService.kt
package com.server.computerscience.chatbot.service.implement

import com.server.computerscience.chatbot.dto.request.ChatMessageDto
import org.springframework.cache.CacheManager
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.security.web.access.WebInvocationPrivilegeEvaluator
import org.springframework.stereotype.Service
import java.util.*

@Service
class ChatCacheService(
    private val cacheManager: CacheManager,
    private val privilegeEvaluator: WebInvocationPrivilegeEvaluator
) {
    val DEFAULT_USED_COUNT: Int = 0;

    /**
     * 이미 저장되어있는 대화 목록이 있을 경우 기존 목록 반환, 아니라면 비어있는 목록을 반환
     */
    private fun getChatMessages(userId: String): List<ChatMessageDto?> {
        val cache = cacheManager.getCache(CHAT_MESSAGE)
        if (cache != null) {
            val chatMessages = cache.get(userId, List::class.java)
            //filterIsInstance는 리스트내에서 특정타입의 요소들만 필터링 한다 -> as 보다 타입 안전성을 확보하면서 제네릭 타입을 활용 가능하다.
            // reified를 사용하면 필터링없이 바로 반환 가능하지만, cache.get는 스프링 내장함수고 일반의 T를 반환하다.
            return chatMessages?.filterIsInstance<ChatMessageDto>() ?: LinkedList()
        }
        return LinkedList()
    }

    @CachePut(value = [CHAT_MESSAGE], key = "#userId")
    fun saveChatMessage(
        userId: String,
        chatMessageDto: ChatMessageDto,
        maxMessageSize: Int
    ): List<ChatMessageDto> {
        val chatMessages: MutableList<ChatMessageDto> = getChatMessages(userId).filterNotNull().toMutableList()
        chatMessages.add(chatMessageDto)
        if (chatMessages.size > maxMessageSize) {
            chatMessages.removeAt(0)
        }
        return chatMessages.toList()
    }

    //남은 이용 횟수를 가져오는 메소드
    fun getUsedChance(userId: String): Int {
        val cache = cacheManager.getCache(CHAT_USED_CHANCE)
        return cache?.get(userId)?.get() as? Int ?: DEFAULT_USED_COUNT
    }

    //사용 횟수 증가 메소드
    @CachePut(value = [CHAT_USED_CHANCE], key = "#userId")
    fun increaseUsedChance(userId: String): Int {
        val usedChance = getUsedChance(userId)
        return usedChance.inc()
    }

    /**
    매 시간마다 모든 사용자에 대한 사용 횟수 캐시를 삭제하는 메서드
     */
    @Scheduled(cron = "0 0 * * * *") // 매 시간 정각에 실행
    @CacheEvict(value = [CHAT_USED_CHANCE], allEntries = true)
    open fun clearAllUsedChance() {
    }

    companion object {
        const val CHAT_MESSAGE = "chatMessages"
        const val CHAT_USED_CHANCE = "chatRemainChance";
    }
}
